common/app/common/pages_helpers.cljc

```diff
@@ -113,3 +166,55 @@
                      (lazy-seq (loopfn (rest ids))))))]
     (loopfn (:shapes root))))
 
+(defn clone-object
+  "Gets a copy of the object and all its children, with new ids
+  and with the parent-children links correctly set. Admits functions
+  to make more transformations to the cloned objects and the
+  original ones.
+
+  Returns the cloned object, the list of all new objects (including
+  the cloned one), and possibly a list of original objects modified."
+
+  ([object parent-id objects xf-new-object]
+   (clone-object object parent-id objects xf-new-object identity))
+
+  ([object parent-id objects xf-new-object xf-original-object]
```

Comment:
The `xf` prefix is used in `transducers` but this functions seems like simple function callbacks

Reply:
Changed to update-xxx